src/metabase/mbql/normalize.cljc

```diff
@@ -639,25 +639,40 @@
 
 (defn- canonicalize-mbql-clauses
   "Walk an `mbql-query` an canonicalize non-top-level clauses like `:fk->`."
-  [mbql-query]
-  (walk/prewalk
-   (fn [x]
-     (cond
-       (map? x)
-       (m/map-vals canonicalize-mbql-clauses x)
-
-       (not (mbql-clause? x))
-       x
-
-       :else
-       (try
-         (canonicalize-mbql-clause x)
-         (catch #?(:clj Throwable :cljs js/Error) e
-           (log/error (i18n/tru "Invalid clause:") x)
-           (throw (ex-info (i18n/tru "Invalid MBQL clause: {0}" (ex-message e))
-                           {:clause x}
-                           e))))))
-   mbql-query))
+  [form]
+  (cond
+    (record? form)
+    (reduce-kv (fn [r k x] (assoc r k (canonicalize-mbql-clauses x))) form form)
+
+    (map? form)
+    (m/map-vals canonicalize-mbql-clauses form)
```

Comment:
Not really asking you to change this since it's existing code, but just wanted to point out that instead of using `m/map-vals` we should probably just use `update-vals` instead. `m/map-vals` isn't really needed now that there's an official Clojure core function that does the same thing. I have a PR to enforce this with a Kondo linter https://github.com/metabase/metabase/pull/32423 that I'm going to hopefully get green at some point

Reply:
Yes, I was considering this change. I'll do it in a separate PR.